### ISSUE TITLE
Design Morpheme Composition Evaluation Metrics

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#47]
## Problem

After training morpheme-aware embeddings (Issue #45), we need specific metrics to evaluate whether the compositional structure is working correctly.

Standard embedding evaluations (cosine similarity, analogy tasks) don't test the key claims of morpheme-aware embeddings:
1. **Compositional generalization**: Can we compose unseen word combinations?
2. **Affix semantics**: Do affixes transform meaning correctly?
3. **Separation of concerns**: Are semantic and grammatical components independent?

### 1. Compositional Generalization Test

**Task**: Evaluate embeddings on unseen word combinations

**Test set**:
- Hold out 10% of root+affix combinations during training
- Never train on "hund+ej" but train on "hund" and "kat+ej"
- At eval: Compose "hundejo" and test in context

**Metrics**:
- Context appropriateness: Does "hundejo" fit in "La hundejo estas granda"?
- Similarity to known compositions: Is "hundejo" ≈ "katejo" (both places)?

### 2. Affix Transformation Test

**Task**: Verify affixes transform root meanings correctly

**Examples**:
```
mal+bon should be similar to mal+bel (both opposites)
mal+bon should be dissimilar to bon (opposite)
hund+ej should be similar to kat+ej (both places)
hund+et should be different from hund+eg (small vs large)
```

**Metrics**:
- Affix consistency: cos(mal+bon, mal+bel) > 0.7
- Semantic difference: cos(mal+bon, bon) < 0.5
- Cross-root patterns: cos(hund+ej, kat+ej) > 0.6

### 3. Semantic-Grammatical Independence Test

**Task**: Verify grammatical changes don't affect semantics

**Examples**:
```
hundo (nominative) vs hundon (accusative) → semantic component identical
hundoj (plural) vs hundo (singular) → semantic component similar
```

**Metrics**:
- Extract semantic and grammatical components separately
- Semantic similarity: cos(hundo_sem, hundon_sem) > 0.95
- Grammatical difference: cos(hundo_gram, hundon_gram) < 0.8

### 4. Root Similarity Preservation

**Task**: Verify roots maintain semantic relationships

**Test set**: Use existing Tatoeba EN-EO pairs as oracle
```
If translate(dog, EO) = hundo and translate(cat, EO) = kato
Then: cos(hund_root, kat_root) should match cos(dog, cat) in English embeddings
```

**Metrics**:
- Correlation with English similarities: Pearson r > 0.6
- Hierarchy preservation: Animals cluster together, tools cluster together

### 5. Downstream Task Performance

**Task**: Measure impact on actual retrieval/QA

**Baselines**:
- Whole-word embeddings (current TreeLSTM)
- Simple root-only embeddings (no affixes)
- Morpheme-aware embeddings (ours)

**Metrics**:
- Retrieval precision@k on M1 corpus
- QA accuracy on Esperanto test set
- Generalization to rare root combinations

### Phase 1: Create Test Sets (2-3 hours)
1. Extract compositional test cases from M1 corpus
2. Hold out 10% of combinations
3. Create affix transformation pairs
4. Build semantic-grammatical test pairs

### Phase 2: Implement Metrics (2-3 hours)
1. Compositional similarity tests
2. Affix consistency checks
3. Component independence tests
4. Downstream task wrappers

### Phase 3: Run Evaluation (30 min)
1. Load trained embeddings from Issue #45
2. Run all metrics
3. Compare to baselines
4. Generate evaluation report

## Success Criteria

**Good embeddings should achieve**:
- Compositional generalization: >70% context fit for unseen combinations
- Affix consistency: Pearson r > 0.7 across affix types
- Semantic-grammatical independence: >0.95 semantic preservation
- Root similarity: Pearson r > 0.6 vs English oracle
- Downstream performance: Equal or better than whole-word baseline

## Related Issues

- Issue #43: MorphemeAwareEmbedding architecture
- Issue #45: Training morpheme embeddings
- Issue #46: OOV root handling (affects generalization tests)

## Priority

**High (P1)** - Should implement alongside training (Issue #45) to validate approach

## Next Steps

1. Create test set extraction script
2. Implement evaluation metrics
3. Add to training script as validation step
4. Use results to tune training objectives
